src/host/sparse.jl

```diff
@@ -0,0 +1,79 @@
+## General Sparse Matrix
+
+Base.size(A::AbstractGPUSparseMatrix) = (A.m, A.n)
+
+SparseArrays.nonzeros(A::AbstractGPUSparseMatrix) = A.nzval
+SparseArrays.getnzval(A::AbstractGPUSparseMatrix) = nonzeros(A)
+SparseArrays.nnz(A::AbstractGPUSparseMatrix) = length(nzval(A))
+
+function LinearAlgebra.rmul!(A::AbstractGPUSparseMatrix, x::Number)
+    rmul!(SparseArrays.getnzval(A), x)
+    return A
+end
+
+function LinearAlgebra.lmul!(x::Number, A::AbstractGPUSparseMatrix)
+    lmul!(x, SparseArrays.getnzval(A))
+    return A
+end
+
+## CSC Matrix
+
+SparseArrays.getcolptr(A::AbstractGPUSparseMatrixCSC) = A.colptr
+SparseArrays.rowvals(A::AbstractGPUSparseMatrixCSC) = A.rowval
+SparseArrays.getrowval(A::AbstractGPUSparseMatrixCSC) = rowvals(A)
+# SparseArrays.nzrange(A::AbstractGPUSparseMatrixCSC, col::Integer) = getcolptr(A)[col]:(getcolptr(A)[col+1]-1) # TODO: this uses scalar indexing
+
+function _goodbuffers_csc(m, n, colptr, rowval, nzval)
+    return (length(colptr) == n + 1 && length(rowval) == length(nzval))
+    # TODO: also add the condition that colptr[end] - 1 == length(nzval) (allowscalar?)
+end
+
+@inline function LinearAlgebra.mul!(C::AbstractGPUVector, A::AnyGPUSparseMatrixCSC, B::AbstractGPUVector, α::Number, β::Number)
+    return LinearAlgebra.generic_matvecmul!(C, LinearAlgebra.wrapper_char(A), LinearAlgebra._unwrap(A), B, LinearAlgebra.MulAddMul(α, β))
+end
+
+@inline function LinearAlgebra.generic_matvecmul!(C::AbstractGPUVector, tA, A::AbstractGPUSparseMatrixCSC, B::AbstractGPUVector, _add::LinearAlgebra.MulAddMul)
+    return SparseArrays.spdensemul!(C, tA, 'N', A, B, _add)
+end
+
+Base.@constprop :aggressive function SparseArrays.spdensemul!(C::AbstractGPUVecOrMat, tA, tB, A::AbstractGPUSparseMatrixCSC, B::AbstractGPUVecOrMat, _add::LinearAlgebra.MulAddMul)
+    if tA == 'N'
+        return _spmatmul!(C, A, wrap(B, tB), _add.alpha, _add.beta)
+    else
+        throw(ArgumentError("tA different from 'N' not yet supported"))
+    end
+end
+
+function _spmatmul!(C::AbstractGPUVecOrMat, A::AbstractGPUSparseMatrixCSC, B::AbstractGPUVecOrMat, α::Number, β::Number)
+    size(A, 2) == size(B, 1) ||
+        throw(DimensionMismatch("second dimension of A, $(size(A,2)), does not match the first dimension of B, $(size(B,1))"))
+    size(A, 1) == size(C, 1) ||
+        throw(DimensionMismatch("first dimension of A, $(size(A,1)), does not match the first dimension of C, $(size(C,1))"))
+    size(B, 2) == size(C, 2) ||
+        throw(DimensionMismatch("second dimension of B, $(size(B,2)), does not match the second dimension of C, $(size(C,2))"))
+    
+    A_colptr = getcolptr(A)
+    A_rowval = rowvals(A)
+    A_nzval = getnzval(A)
```

Comment:
Why deconstruct these? Wouldn't it be cleaner to expect the back-end to also provide a device-type for sparse arrays, so that you can pass it directly?

Reply:
What do you mean exactly? To do the multiplication I need to extract these data, which are essentially backend-specific vectors.

Reply:
If the kernel still calls `getrowptr` and the others, I don't see the difference then. Instead of extracting them outside the kernel it would be done inside. Any CSC sparse matrix should have those fields by definition.

Reply:
I tried but it reruns scalar indexing when I do `getcolptr(A)[col]:(getcolptr(A)[col+1]-1)`, which doesn't occur when I do `A_colptr[col]:(A_colptr[col+1]-1)`

Reply:
Scalar indexing cannot happen from a kernel, so I'm confused about the context here.

Reply:
This is what I get with the following function

```julia
function _spmatmul!(C::AbstractGPUVecOrMat, A::AbstractGPUSparseMatrixCSC, B::AbstractGPUVecOrMat, α::Number, β::Number)
    backend_C = KernelAbstractions.get_backend(C)
    backend_A = KernelAbstractions.get_backend(A)
    backend_B = KernelAbstractions.get_backend(B)

    backend_A == backend_B == backend_C || throw(ArgumentError("All arrays must be on the same backend"))

    @kernel function kernel_spmatmul_N!(C, @Const(A), @Const(B))
        k, col = @index(Global, NTuple)

        Bi, Bj = col, k

        @inbounds axj = B[Bi, Bj] * α
        @inbounds for j in getcolptr(A)[col]:(getcolptr(A)[col+1]-1) # nzrange(A, col)
            KernelAbstractions.@atomic C[getrowval(A)[j], k] += getnzval(A)[j] * axj
        end
    end

    β != one(β) && LinearAlgebra._rmul_or_fill!(C, β)

    kernel! = kernel_spmatmul_N!(backend_A)
    kernel!(C, A, B; ndrange=(size(C, 2), size(A, 2)))

    return C
end
```

```julia
ERROR: TaskFailedException

    nested task error: Scalar indexing is disallowed.
    Invocation of getindex resulted in scalar indexing of a GPU array.
    This is typically caused by calling an iterating implementation of a method.
    Such implementations *do not* execute on the GPU, but very slowly on the CPU,
    and therefore should be avoided.

    If you want to allow scalar iteration, use `allowscalar` or `@allowscalar`
    to enable scalar iteration globally or for the operations in question.
    Stacktrace:
      [1] error(s::String)
        @ Base ./error.jl:35
      [2] errorscalar(op::String)
        @ GPUArraysCore ~/GitHub/GPUArrays.jl/lib/GPUArraysCore/src/GPUArraysCore.jl:170
      [3] _assertscalar(op::String, behavior::GPUArraysCore.ScalarIndexing)
        @ GPUArraysCore ~/GitHub/GPUArrays.jl/lib/GPUArraysCore/src/GPUArraysCore.jl:143
      [4] assertscalar(op::String)
        @ GPUArraysCore ~/GitHub/GPUArrays.jl/lib/GPUArraysCore/src/GPUArraysCore.jl:131
      [5] getindex
        @ ~/GitHub/GPUArrays.jl/src/host/indexing.jl:50 [inlined]
      [6] macro expansion
        @ ~/GitHub/GPUArrays.jl/src/host/sparse.jl:157 [inlined]
      [7] cpu_kernel_spmatmul_N!
        @ ~/.julia/packages/KernelAbstractions/0r40T/src/macros.jl:293 [inlined]
      [8] (::GPUArrays.var"#cpu_kernel_spmatmul_N!#190"{…})(__ctx__::KernelAbstractions.CompilerMetadata{…}, C::JLArrays.JLDeviceArray{…}, A::JLSparseMatrixCSC{…}, B::JLArrays.JLDeviceArray{…})
        @ GPUArrays ./none:0
      [9] __thread_run(tid::Int64, len::Int64, rem::Int64, obj::KernelAbstractions.Kernel{…}, ndrange::Tuple{…}, iterspace::KernelAbstractions.NDIteration.NDRange{…}, args::Tuple{…}, dynamic::KernelAbstractions.NDIteration.DynamicCheck)
        @ KernelAbstractions ~/.julia/packages/KernelAbstractions/0r40T/src/cpu.jl:142
     [10] (::KernelAbstractions.var"#20#23"{…})()
        @ KernelAbstractions ~/.julia/packages/KernelAbstractions/0r40T/src/cpu.jl:117

...and 15 more exceptions
```

Referencing to the line of `getcolptr`.

Reply:
Its seems from here

```julia
(::GPUArrays.var"#cpu_kernel_spmatmul_N!#190"{Bool})(__ctx__::KernelAbstractions.CompilerMetadata{KernelAbstractions.NDIteration.DynamicSize, KernelAbstractions.NDIteration.DynamicCheck, CartesianIndex{2}, CartesianIndices{2, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}}, KernelAbstractions.NDIteration.NDRange{2, KernelAbstractions.NDIteration.DynamicSize, KernelAbstractions.NDIteration.DynamicSize, CartesianIndices{2, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}}, CartesianIndices{2, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}}}}, C::JLArrays.JLDeviceArray{Float32, 1}, A::JLSparseMatrixCSC{Float32, Int64}, B::JLArrays.JLDeviceArray{Float32, 1})
```

that the compiler doesn't know about the `JLDeviceArray`s inside `JLSparseMatrixCSC`.

Reply:
`gpu_test_kernel(::KernelAbstractions.CompilerMetadata{…}, ::Vector{…}, ::MyStruct{…})`

You're passing a CPU vector to this kernel...

> Moreover, I think we should still support `JLArrays`, in order to have working tests in GPUArrays.jl, right?

That's correct; I don't think I suggested otherwise?

Reply:
> You're passing a CPU vector to this kernel

Oh, that's true. Now, by using `y = similar(x.data)` I still get the following error

```julia
LoadError: GPU compilation of MethodInstance for gpu_test_kernel(::KernelAbstractions.CompilerMetadata{KernelAbstractions.NDIteration.DynamicSize, 
KernelAbstractions.NDIteration.DynamicCheck, Nothing, CartesianIndices{1, Tuple{Base.OneTo{Int64}}}, 
KernelAbstractions.NDIteration.NDRange{1, KernelAbstractions.NDIteration.DynamicSize, 
KernelAbstractions.NDIteration.DynamicSize, CartesianIndices{1, Tuple{Base.OneTo{Int64}}}, CartesianIndices{1, 
Tuple{Base.OneTo{Int64}}}}}, ::CuDeviceVector{Float32, 1}, ::MyStruct{Float32}) failed
KernelError: passing non-bitstype argument

Argument 4 to your kernel function is of type MyStruct{Float32}, which is not a bitstype:
  .data is of type CuArray{Float32, 1} which is not isbits.
    .data is of type DataRef{CUDA.Managed{M}} where M which is not isbits.
      .rc is of type GPUArrays.RefCounted{CUDA.Managed{M}} where M which is not isbits.
        .obj is of type CUDA.Managed which is not isbits.
          .mem is of type Any which is not isbits.
          .stream is of type CuStream which is not isbits.
            .ctx is of type Union{Nothing, CuContext} which is not isbits.
        .finalizer is of type Any which is not isbits.
        .count is of type Base.Threads.Atomic{Int64} which is not isbits.


Only bitstypes, which are "plain data" types that are immutable
and contain no references to other values, can be used in GPU kernels.
For more information, see the `Base.isbitstype` function.

Stacktrace:
  [1] check_invocation(job::GPUCompiler.CompilerJob)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/GnbhK/src/validation.jl:92
  [2] macro expansion
    @ ~/.julia/packages/GPUCompiler/GnbhK/src/driver.jl:92 [inlined]
  [3] macro expansion
    @ ~/.julia/packages/TimerOutputs/6KVfH/src/TimerOutput.jl:253 [inlined]
  [4] codegen(output::Symbol, job::GPUCompiler.CompilerJob; toplevel::Bool, libraries::Bool, optimize::Bool, cleanup::Bool, validate::Bool, strip::Bool, only_entry::Bool, parent_job::Nothing)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/GnbhK/src/driver.jl:90
  [5] codegen
    @ ~/.julia/packages/GPUCompiler/GnbhK/src/driver.jl:82 [inlined]
  [6] compile(target::Symbol, job::GPUCompiler.CompilerJob; kwargs::@Kwargs{})
    @ GPUCompiler ~/.julia/packages/GPUCompiler/GnbhK/src/driver.jl:79
  [7] compile
    @ ~/.julia/packages/GPUCompiler/GnbhK/src/driver.jl:74 [inlined]
  [8] #1145
    @ ~/.julia/packages/CUDA/gqdCz/src/compiler/compilation.jl:250 [inlined]
  [9] JuliaContext(f::CUDA.var"#1145#1148"{GPUCompiler.CompilerJob{GPUCompiler.PTXCompilerTarget, CUDA.CUDACompilerParams}}; kwargs::@Kwargs{})
    @ GPUCompiler ~/.julia/packages/GPUCompiler/GnbhK/src/driver.jl:34
 [10] JuliaContext(f::Function)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/GnbhK/src/driver.jl:25
 [11] compile(job::GPUCompiler.CompilerJob)
    @ CUDA ~/.julia/packages/CUDA/gqdCz/src/compiler/compilation.jl:249
 [12] actual_compilation(cache::Dict{Any, CuFunction}, src::Core.MethodInstance, world::UInt64, cfg::GPUCompiler.CompilerConfig{GPUCompiler.PTXCompilerTarget, CUDA.CUDACompilerParams}, compiler::typeof(CUDA.compile), linker::typeof(CUDA.link))
    @ GPUCompiler ~/.julia/packages/GPUCompiler/GnbhK/src/execution.jl:237
 [13] cached_compilation(cache::Dict{Any, CuFunction}, src::Core.MethodInstance, cfg::GPUCompiler.CompilerConfig{GPUCompiler.PTXCompilerTarget, CUDA.CUDACompilerParams}, compiler::Function, linker::Function)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/GnbhK/src/execution.jl:151
 [14] macro expansion
    @ ~/.julia/packages/CUDA/gqdCz/src/compiler/execution.jl:380 [inlined]
 [15] macro expansion
    @ ./lock.jl:273 [inlined]
 [16] cufunction(f::typeof(gpu_test_kernel), tt::Type{Tuple{KernelAbstractions.CompilerMetadata{KernelAbstractions.NDIteration.DynamicSize, KernelAbstractions.NDIteration.DynamicCheck, Nothing, CartesianIndices{1, Tuple{Base.OneTo{Int64}}}, KernelAbstractions.NDIteration.NDRange{1, KernelAbstractions.NDIteration.DynamicSize, KernelAbstractions.NDIteration.DynamicSize, CartesianIndices{1, Tuple{Base.OneTo{Int64}}}, CartesianIndices{1, Tuple{Base.OneTo{Int64}}}}}, CuDeviceVector{Float32, 1}, MyStruct{Float32}}}; kwargs::@Kwargs{always_inline::Bool, maxthreads::Nothing})
    @ CUDA ~/.julia/packages/CUDA/gqdCz/src/compiler/execution.jl:375
 [17] macro expansion
    @ ~/.julia/packages/CUDA/gqdCz/src/compiler/execution.jl:112 [inlined]
 [18] (::KernelAbstractions.Kernel{CUDABackend, KernelAbstractions.NDIteration.DynamicSize, KernelAbstractions.NDIteration.DynamicSize, typeof(gpu_test_kernel)})(::CuArray{Float32, 1, CUDA.DeviceMemory}, ::Vararg{Any}; ndrange::Tuple{Int64}, workgroupsize::Nothing)
    @ CUDA.CUDAKernels ~/.julia/packages/CUDA/gqdCz/src/CUDAKernels.jl:103
```

> I don't think I suggested otherwise?

> You're using the KA CPU back-end because you didn't register a back-end a la

I thought that the problem was because I was just using `JLArray`s. I have already defined

```julia
KernelAbstractions.get_backend(x::MyStruct) = KernelAbstractions.get_backend(x.data)
```

Reply:
Ok, it should be fixed now. Thanks for your help.